crates/accelerate/src/high_level_synthesis.rs

```diff
@@ -171,82 +213,807 @@ impl QubitTracker {
+    let mut output_qubits = input_qubits.to_vec();
+
+    // The "inverse" map from the global qubits to the output circuit's qubits.
+    // This map may be extended if additional auxiliary qubits get used.
+    let mut global_to_local: HashMap<usize, usize> =
+        HashMap::from_iter(output_qubits.iter().enumerate().map(|(i, j)| (*j, i)));
+
+    for inst in input_circuit.iter() {
+        // op's qubits as viewed globally
+        let op_qubits = input_circuit
+            .get_qargs(inst.qubits)
+            .iter()
+            .map(|q| input_qubits[q.index()])
+            .collect::<Vec<usize>>();
+
+        // Start by handling special operations.
+        // In the future, we can also consider other possible optimizations, e.g.:
+        //   - improved qubit tracking after a SWAP gate
+        //   - automatically simplify control gates with control at 0.
+        if ["id", "delay", "barrier"].contains(&inst.op.name()) {
+            output_circuit.push(py, inst.clone())?;
+            // tracker is not updated, these are no-ops
+            continue;
+        }
+
+        if inst.op.name() == "reset" {
+            output_circuit.push(py, inst.clone())?;
+            tracker.borrow_mut().set_clean(op_qubits);
+            continue;
+        }
+
+        // Check if synthesis for this operation can be skipped
+        let op_qargs: Vec<Qubit> = op_qubits.iter().map(|q| Qubit(*q as u32)).collect();
+        if definitely_skip_op(py, data, &inst.op, &op_qargs) {
+            output_circuit.push(py, inst.clone())?;
+            tracker.borrow_mut().set_dirty(op_qubits);
+            continue;
+        }
+
+        // Recursively handle control-flow.
+        // Currently we do not allow subcircuits within the control flow to use auxiliary qubits
+        // and mark all the usable qubits as dirty. This is done in order to avoid complications
+        // that different subcircuits may choose to use different auxiliary global qubits, and to
+        // avoid complications related to tracking qubit status for while- loops.
+        // In the future, this handling can potentially be improved.
+        if inst.op.control_flow() {
+            let quantum_circuit_cls = QUANTUM_CIRCUIT.get_bound(py);
+            if let OperationRef::Instruction(py_inst) = inst.op.view() {
+                let old_blocks_as_bound_obj = py_inst.instruction.bind(py);
+
+                // old_blocks_py keeps the orignal QuantumCircuit's appearing within control-flow ops
+                // new_blocks_py keeps the recursively synthesized circuits
+                let old_blocks_py = old_blocks_as_bound_obj.getattr(intern!(py, "blocks"))?;
+                let old_blocks_py = old_blocks_py.downcast::<PyTuple>()?;
+                let mut new_blocks_py: Vec<Bound<PyAny>> = Vec::with_capacity(old_blocks_py.len());
+
+                // We do not allow using any additional qubits outside of the block.
+                let block_tracker = tracker.clone();
```

Comment:
I'm not sure this will work as intended. When you call clone on a `Bound` instance, what it does is increast the reference count in Python. But it is still pointing towards the same object. If you intended to modify the original one, then this should be fine. But if your intention was to make a clone of the original object and then modify it, this might not be doing that.

Reply:
Great catch! This could have been a problem indeed, but I think it's gone now, now that `tracker` is a `QubitTracker` and not a `Bound<QubitTracker>`.